### DISCUSSION TITLE
Package Labels

### DISCUSSION THREAD

**Discussion by IncredibleX** [abuzuhri/Amazon-SP-API-CSharp#152] [Q&A]
Hi,

i like to know how to download the PDF Label for PartneredSmallParcelData after i confirmed via SPAPI.

Any idea?

Thanx
Tobias

**Comment by abuzuhri** [comment]
no idea which report you need to call

**Reply by IncredibleX** [reply]
I think it is not possible.

**Comment by IncredibleX** [comment]
I don’t think, it is a report.

I think it is not possible download the label. My solution is to create an own report
similar to amazons report.

I sent a test package to amazon and this did work. So i generate the labels myself with given data.


Am 22. Apr. 2022, 06:13 +0200 schrieb Tareq Abuzuhri ***@***.***>:

**Comment by IncredibleX** [comment]
My solution:
My code snippet for creating a shipmentplan and request the necessary data to generate an own PDF Label for DHL Shipment.
Perhaps this helps somebody to generate the labels. This code works is now daily in use for FBAInbound delivery.

`  /// <summary>
            /// Anlieferplan
            /// </summary>
            public string CreateShipmentPlan(string Shipmentname, string Sendershortname, double weigth, double pkgWidth, double pkgHeight, double pkgLength)
            {
                oTools = new ToolsAndHelpers.Tools(exSettings);
      
                eXODALib.ItemLib.Items oItems = new ItemLib.Items(exSettings);
                eXODALib.ItemLib.Item oItem = new ItemLib.Item(exSettings);
                //FikaAmazonAPI.AmazonSpApiSDK.Models.FulfillmentInbound.GetTransportDetailsResult oGetTransportDetailsResult = amazonConnection.FulFillmentInbound.GetTransportDetails(stestShipmentID);

                //this.PutTransportDetails(stestShipmentID);
                //FikaAmazonAPI.AmazonSpApiSDK.Models.FulfillmentInbound.GetShipmentsResult oGetShipmentsResult = this.GetShipmentsResult(stestShipmentID);
                //this.GetLabels(stestShipmentID);
                //return;


                FikaAmazonAPI.AmazonSpApiSDK.Models.FulfillmentInbound.CreateInboundShipmentPlanRequest oCreateInboundShipmentPlanRequest = new FikaAmazonAPI.AmazonSpApiSDK.Models.FulfillmentInbound.CreateInboundShipmentPlanRequest();

                oCreateInboundShipmentPlanRequest.ShipToCountryCode = "DE";
                oCreateInboundShipmentPlanRequest.LabelPrepPreference = FikaAmazonAPI.AmazonSpApiSDK.Models.FulfillmentInbound.LabelPrepPreference.SELLERLABEL;


                eXODALib.AddressLib.AddressOrganisation oAdrOrga = new AddressLib.AddressOrganisation();
                oAdrOrga.Initialize(exSettings);
                eXODALib.AddressLib.Address oAdr = new AddressLib.Address();
                oAdr = oAdrOrga.Open(Sendershortname);
                oCreateInboundShipmentPlanRequest.ShipFromAddress = new FikaAmazonAPI.AmazonSpApiSDK.Models.FulfillmentInbound.Address();
                oCreateInboundShipmentPlanRequest.ShipFromAddress.AddressLine1 = oAdr.Name2;
                oCreateInboundShipmentPlanRequest.ShipFromAddress.AddressLine2 = oAdr.Street;
                oCreateInboundShipmentPlanRequest.ShipFromAddress.City = oAdr.City;
                oCreateInboundShipmentPlanRequest.ShipFromAddress.CountryCode = "DE";
                oCreateInboundShipmentPlanRequest.ShipFromAddress.PostalCode = oAdr.Postcode;
                oCreateInboundShipmentPlanRequest.ShipFromAddress.Name = oAdr.Name1;




                oCreateInboundShipmentPlanRequest.InboundShipmentPlanRequestItems = new FikaAmazonAPI.AmazonSpApiSDK.Models.FulfillmentInbound.InboundShipmentPlanRequestItemList();
                FikaAmazonAPI.AmazonSpApiSDK.Models.FulfillmentInbound.InboundShipmentPlanRequestItem oInboundShipmentPlanRequestItem = new FikaAmazonAPI.AmazonSpApiSDK.Models.FulfillmentInbound.InboundShipmentPlanRequestItem();

                string sSQL = "Select * from AmazonFBASendungen where Menge <> 0 and  Sendungsname = '" + Shipmentname + "' and Mandant = '" + this.exSettings.eXODAClientname + "'";
                this.exSettings.HS4Dataset.Open(HS4Software.DatabaseType.ExistingConnection, "", sSQL, "shipmentsitems", "ID", true, false);
                this.exSettings.HS4Dataset.SetAutoIncrementColumn("ID", "shipmentsitems");


                foreach (DataRow row in this.exSettings.HS4Dataset.HSDataSet.Tables["shipmentsitems"].Rows)
                {
                    oItem = oItems.LoadItem(Convert.ToInt64(row["SKU"]));
                    oInboundShipmentPlanRequestItem = new FikaAmazonAPI.AmazonSpApiSDK.Models.FulfillmentInbound.InboundShipmentPlanRequestItem();
                    oInboundShipmentPlanRequestItem.SellerSKU = oItem.ItemID.ToString();
                    oInboundShipmentPlanRequestItem.Quantity = Convert.ToInt16(row["Menge"]);
                    oInboundShipmentPlanRequestItem.Condition = FikaAmazonAPI.AmazonSpApiSDK.Models.FulfillmentInbound.Condition.NewItem;
                    //oInboundShipmentPlanRequestItem.ASIN = oItem.AmazonASIN;

                    oCreateInboundShipmentPlanRequest.InboundShipmentPlanRequestItems.Add(oInboundShipmentPlanRequestItem);
                }

                FikaAmazonAPI.AmazonSpApiSDK.Models.FulfillmentInbound.CreateInboundShipmentPlanResult oCreateInboundShipmentPlanResult = amazonConnection.FulFillmentInbound.CreateInboundShipmentPlan(oCreateInboundShipmentPlanRequest);


                string sNewShipmentname = Shipmentname;

                int iPlanCount = oCreateInboundShipmentPlanResult.InboundShipmentPlans.Count;
                for (int iPlan = 0; iPlan < iPlanCount; iPlan++)
                {
                    string sShipmentID = oCreateInboundShipmentPlanResult.InboundShipmentPlans[iPlan].ShipmentId;
                    if (iPlanCount > 1)
                        sNewShipmentname = Shipmentname + "_" + oCreateInboundShipmentPlanResult.InboundShipmentPlans[iPlan].DestinationFulfillmentCenterId + "_" + Convert.ToString(iPlan + 1) + "_von_" + iPlanCount.ToString();


                    //this.exSettings.HS4Dataset.Open(HS4Software.DatabaseType.ExistingConnection, "", sSQL, "shipmentsitems", "ID", true, false);

                    FikaAmazonAPI.AmazonSpApiSDK.Models.FulfillmentInbound.InboundShipmentRequest oInboundShipmentRequest = new FikaAmazonAPI.AmazonSpApiSDK.Models.FulfillmentInbound.InboundShipmentRequest();

                    double dPackageWeight = 0;
                    oInboundShipmentRequest.InboundShipmentItems = new FikaAmazonAPI.AmazonSpApiSDK.Models.FulfillmentInbound.InboundShipmentItemList();
                    foreach (DataRow row in this.exSettings.HS4Dataset.HSDataSet.Tables["shipmentsitems"].Rows)
                    {


                        oInboundShipmentRequest.InboundShipmentHeader = new FikaAmazonAPI.AmazonSpApiSDK.Models.FulfillmentInbound.InboundShipmentHeader();
                        oInboundShipmentRequest.InboundShipmentHeader.ShipFromAddress = oCreateInboundShipmentPlanRequest.ShipFromAddress;
                        oInboundShipmentRequest.InboundShipmentHeader.ShipmentName = Shipmentname;
                        oInboundShipmentRequest.InboundShipmentHeader.DestinationFulfillmentCenterId = oCreateInboundShipmentPlanResult.InboundShipmentPlans[iPlan].DestinationFulfillmentCenterId;

                        oInboundShipmentRequest.InboundShipmentHeader.AreCasesRequired = false;
                        oInboundShipmentRequest.InboundShipmentHeader.IntendedBoxContentsSource = FikaAmazonAPI.AmazonSpApiSDK.Models.FulfillmentInbound.IntendedBoxContentsSource.FEED;


                        FikaAmazonAPI.AmazonSpApiSDK.Models.FulfillmentInbound.InboundShipmentItem oInboundShipmentItem = new FikaAmazonAPI.AmazonSpApiSDK.Models.FulfillmentInbound.InboundShipmentItem();


                        oItem = oItems.LoadItem(Convert.ToInt64(row["SKU"]));
                        oInboundShipmentItem.SellerSKU = oItem.ItemID.ToString();
                        oInboundShipmentItem.QuantityShipped = Convert.ToInt16(row["Menge"]);
                        dPackageWeight = dPackageWeight + (oItem.Weight * Convert.ToInt16(row["Menge"]));

                        oInboundShipmentRequest.InboundShipmentItems.Add(oInboundShipmentItem);


                    }
                    FikaAmazonAPI.AmazonSpApiSDK.Models.FulfillmentInbound.InboundShipmentResult oInboundShipmentResult = amazonConnection.FulFillmentInbound.CreateInboundShipment(oCreateInboundShipmentPlanResult.InboundShipmentPlans[iPlan].ShipmentId, oInboundShipmentRequest);

                    FikaAmazonAPI.AmazonSpApiSDK.Models.FulfillmentInbound.PutTransportDetailsRequest oPutTransportDetailsRequest = new FikaAmazonAPI.AmazonSpApiSDK.Models.FulfillmentInbound.PutTransportDetailsRequest();
                    oPutTransportDetailsRequest.IsPartnered = true;
                    oPutTransportDetailsRequest.TransportDetails = new FikaAmazonAPI.AmazonSpApiSDK.Models.FulfillmentInbound.TransportDetailInput();


                    FikaAmazonAPI.AmazonSpApiSDK.Models.FulfillmentInbound.PartneredSmallParcelPackageInput oPackageInputItem = new FikaAmazonAPI.AmazonSpApiSDK.Models.FulfillmentInbound.PartneredSmallParcelPackageInput();
                    oPackageInputItem.Weight = new FikaAmazonAPI.AmazonSpApiSDK.Models.FulfillmentInbound.Weight();
                    oPackageInputItem.Weight.Value = weigth; // dPackageWeight;
                    oPackageInputItem.Weight.Unit = FikaAmazonAPI.AmazonSpApiSDK.Models.FulfillmentInbound.UnitOfWeight.Kilograms;

                    oPackageInputItem.Dimensions = new FikaAmazonAPI.AmazonSpApiSDK.Models.FulfillmentInbound.Dimensions();
                    if (oTools.ClientSetting("FulfillmentInboundStandardDimensionsWidth" + "_" + exSettings.ActiveUser) == null) oTools.SaveClientSetting("FulfillmentInboundStandardDimensionsWidth" + "_" + exSettings.ActiveUser, "50");
                    if (oTools.ClientSetting("FulfillmentInboundStandardDimensionsHeight" + "_" + exSettings.ActiveUser) == null) oTools.SaveClientSetting("FulfillmentInboundStandardDimensionsHeight" + "_" + exSettings.ActiveUser, "50");
                    if (oTools.ClientSetting("FulfillmentInboundStandardDimensionsLength" + "_" + exSettings.ActiveUser) == null) oTools.SaveClientSetting("FulfillmentInboundStandardDimensionsLength" + "_" + exSettings.ActiveUser, "50");

                    oPackageInputItem.Dimensions.Width = pkgWidth;
                    oPackageInputItem.Dimensions.Height = pkgHeight;
                    oPackageInputItem.Dimensions.Length = pkgLength;
                    oPackageInputItem.Dimensions.Unit = FikaAmazonAPI.AmazonSpApiSDK.Models.FulfillmentInbound.UnitOfMeasurement.Centimeters;

                    oPutTransportDetailsRequest.TransportDetails.PartneredSmallParcelData = new FikaAmazonAPI.AmazonSpApiSDK.Models.FulfillmentInbound.PartneredSmallParcelDataInput();
                    oPutTransportDetailsRequest.TransportDetails.PartneredSmallParcelData.PackageList = new FikaAmazonAPI.AmazonSpApiSDK.Models.FulfillmentInbound.PartneredSmallParcelPackageInputList();
                    oPutTransportDetailsRequest.TransportDetails.PartneredSmallParcelData.PackageList.Add(oPackageInputItem);
                    oPutTransportDetailsRequest.ShipmentType = "SP";// FikaAmazonAPI.AmazonSpApiSDK.Models.FulfillmentInbound.ShipmentType.SP;
                    oPutTransportDetailsRequest.TransportDetails.PartneredSmallParcelData.CarrierName = "DHL_STANDARD"; //UNITED_PARCEL_SERVICE_INC



                    amazonConnection.FulFillmentInbound.PutTransportDetails(sShipmentID, oPutTransportDetailsRequest);
                    System.Threading.Thread.Sleep(1000);

                    FikaAmazonAPI.AmazonSpApiSDK.Models.FulfillmentInbound.GetTransportDetailsResult oGetTransportDetailsResult = amazonConnection.FulFillmentInbound.GetTransportDetails(sShipmentID);

                    //while (oGetTransportDetailsResult.TransportContent.TransportResult.TransportStatus !=  FikaAmazonAPI.AmazonSpApiSDK.Models.FulfillmentInbound.TransportStatus.WORKING)
                    //{ 
                    //    Console.WriteLine("Working...");
                    //    if (oGetTransportDetailsResult.TransportContent.TransportResult.ErrorCode != null)
                    //        return oGetTransportDetailsResult.TransportContent.TransportResult.ErrorDescription + " / " + oGetTransportDetailsResult.TransportContent.TransportResult.ErrorCode;

                    //}

                    FikaAmazonAPI.AmazonSpApiSDK.Models.FulfillmentInbound.CommonTransportResult oEstimation = amazonConnection.FulFillmentInbound.EstimateTransport(sShipmentID);
                    System.Threading.Thread.Sleep(1000);

                    //while (oGetTransportDetailsResult.TransportContent.TransportResult.TransportStatus != FikaAmazonAPI.AmazonSpApiSDK.Models.FulfillmentInbound.TransportStatus.WORKING)
                    //{
                    //    Console.WriteLine("Wait for Estimating state...");
                    //    System.Threading.Thread.Sleep(1000);
                    //    if (oGetTransportDetailsResult.TransportContent.TransportResult.TransportStatus == FikaAmazonAPI.AmazonSpApiSDK.Models.FulfillmentInbound.TransportStatus.ESTIMATING)
                    //        return oGetTransportDetailsResult.TransportContent.TransportResult.ErrorDescription + " / " + oGetTransportDetailsResult.TransportContent.TransportResult.ErrorCode;
                    //}

                    FikaAmazonAPI.AmazonSpApiSDK.Models.FulfillmentInbound.CommonTransportResult oConfirmTransport = amazonConnection.FulFillmentInbound.ConfirmTransport(sShipmentID);


                    System.Threading.Thread.Sleep(1000);

                    oInboundShipmentRequest.InboundShipmentHeader.LabelPrepPreference = FikaAmazonAPI.AmazonSpApiSDK.Models.FulfillmentInbound.LabelPrepPreference.SELLERLABEL;
                    oInboundShipmentRequest.InboundShipmentHeader.ShipmentStatus = FikaAmazonAPI.AmazonSpApiSDK.Models.FulfillmentInbound.ShipmentStatus.SHIPPED;
                    FikaAmazonAPI.AmazonSpApiSDK.Models.FulfillmentInbound.InboundShipmentResult oReq = amazonConnection.FulFillmentInbound.UpdateInboundShipment(sShipmentID, oInboundShipmentRequest);

                    oGetTransportDetailsResult = amazonConnection.FulFillmentInbound.GetTransportDetails(sShipmentID);
                    if (oGetTransportDetailsResult.TransportContent.TransportResult.ErrorCode != null)
                        return oGetTransportDetailsResult.TransportContent.TransportResult.ErrorDescription + " / " + oGetTransportDetailsResult.TransportContent.TransportResult.ErrorCode;
                    foreach (DataRow row in this.exSettings.HS4Dataset.HSDataSet.Tables["shipmentsitems"].Rows)
                    {
                        row["SendungsID"] = oGetTransportDetailsResult.TransportContent.TransportHeader.ShipmentId;
                        row["TrackingID"] = oGetTransportDetailsResult.TransportContent.TransportDetails.PartneredSmallParcelData.PackageList[0].TrackingId;
                        row["FulFillmentCenter"] = oCreateInboundShipmentPlanResult.InboundShipmentPlans[iPlan].DestinationFulfillmentCenterId;
                    }

                    this.exSettings.HS4Dataset.Update("shipmentsitems");
                }
                return "OK";
            }`

Here you can see my own designed Report.
<img width="568" alt="Bildschirmfoto 2022-04-22 um 16 12 16" src="https://user-images.githubusercontent.com/38632368/164732155-b8697a8e-eb95-409e-a969-6d606cebfe25.png">
